uips/uip-5.md

```diff
@@ -0,0 +1,118 @@
+    pub source_channel: ChannelId,
+    pub use_compat_address: bool,
+    
+    // New field
+    pub ics20_memo: String,
+}
+```
+
+The `From<Ics20Withdrawal> for pb::FungibleTokenPacketData` impl, which encodes the withdrawal in the ICS-compatible packet data type, SHALL be modified to include the contents of this new field:
+
+```rust
+impl From<Ics20Withdrawal> for pb::FungibleTokenPacketData {
+    fn from(w: Ics20Withdrawal) -> Self {
+        let return_address = match w.use_compat_address {
+            true => w.return_address.compat_encoding(),
+            false => w.return_address.to_string(),
+        };
+
+        pb::FungibleTokenPacketData {
+            amount: w.value().amount.to_string(),
+            denom: w.denom.to_string(),
+            receiver: w.destination_chain_address,
+            sender: return_address,
+            memo: w.ics20_memo,
+        }
+    }
+}
+```
+
+The `ics20_memo` field:
+- MAY contain any valid UTF-8 string
+- MAY be empty
+
+## Rationale
+
+There are two major parts of Packet Forwarding Middleware support:
+
+- Outbound PFM Support (what this UIP supports): the support for sending packets that can be autonomously routed over multiple intermediate chains before reaching their destination.
+- Packet Forwarding Middleware support: the actual forwarding middleware implementation, which runs in the chain's state machine and allows the chain to forward packets sent from other chains.
+
+This UIP implements (1), using a minimal protocol change (adding the configurable memo field). This gives the majority of end-user benefit from PFM support, notably that users can experience superior UX in withdrawing from the Penumbra chain, while avoiding complex protocol development work and assurance. A separate UIP for implementing (2) in the Penumbra IBC stack could be considered in the future.
+
+
+## Backwards Compatibility
+
+This UIP adds an additional field to the Ics20Withdrawal struct, which would be a breaking change and require a protocol upgrade.
+
+## Test Cases
+
+```rust
+#[test]
+fn test_ics20_withdrawal_memo() {
+    let withdrawal = Ics20Withdrawal {
+        amount: Amount::from(1000u64),
+        denom: asset::Metadata::dummy(),
+        destination_chain_address: "cosmos1...".to_string(),
+        return_address: Address::dummy(),
+        timeout_height: IbcHeight::new(1, 1000),
+        timeout_time: 1000000,
+        source_channel: ChannelId::new(0),
+        use_compat_address: false,
+        ics20_memo: "{\"forward\":{...}".to_string(),
+    };
+    
+    let packet_data: FungibleTokenPacketData = withdrawal.into();
+    assert_eq!(packet_data.memo, withdrawal.ics20_memo);
+}
+```
+
+## Security Considerations
+
+This change does not introduce new security considerations at the IBC state machine level. The memo field is already part of the ICS20 specification, this UIP simply opens up the possibility for the user to configure its contents while building a withdrawal transaction.
```

Comment:
Doesn't this open up a security consideration for Penumbra? Currently, we don't have any transaction fields that allow user- (attacker-) controlled lengths; this would be one.

The ICS20 memo field has been used for griefing on other chains. I think this won't be a problem, because Penumbra charges transactions gas for bytes used in the transaction encoding, so an attacker would have to pay to spam the chain, and the gas cost can change / be changed to respond to an attacker. But we should note that in the UIP, I think, so there's a documented rationale.

Reply:
Should we consider adding a maximum size bound to the memo field in stateless checks for Ics20Withdrawal? This isn't specified in the ICS spec as far as I am aware, but a reasonable upper bound could be added with likely few interoperability concerns.

Reply:
I'm in favor of not adding a bound a priori, and just relying on gas ; god knows what legitimate use cases people might come up with

Reply:
The thing about adding a bound is that someone can just run right up to the bound and then do multiple transactions. So pricing by block space seems like the workable solution. However this does suggest that we might need to push forward on dynamic fee pricing sooner rather than later.